### ISSUE TITLE
Sceptre generate command fails with valid yaml

### ISSUE THREAD

**Issue by zaro0508** [Sceptre/sceptre#710]
I have a valid yaml file but when i run sceptre generate it fails with error "An error occured whilst writing the YAML object".  It seems like the problem is the special character `!` in my CFN yaml.  The generate command works if I change `!Ref` to `Ref:`. 

Steps to repo:
```
(py36-default) ❯❯❯ sceptre --version
Sceptre, version 2.1.2

(py36-default) ❯❯❯ cat templates/template.yaml
AWSTemplateFormatVersion: 2010-09-09
Resources:
  MyBucket:
    Type: "AWS::S3::Bucket"
  MyBucketPolicy:
    Type: "AWS::S3::BucketPolicy"
    Properties:
      # sceptre generate fails for this
      Bucket: !Ref MyBucket
      # sceptre generate works for this
      # Bucket:
      #   Ref: MyBucket
      PolicyDocument:
        Version: "2012-10-17"
        Statement:
          - Sid: "AWSCloudTrailAclCheck"
            Effect: "Allow"
            Principal:
              Service: "cloudtrail.amazonaws.com"
            Action: "s3:GetBucketAcl"

(py36-default) ❯❯❯ cat config/prod/test-stack.yaml
template_path: template.yaml
stack_name: test-stack

(py36-default) ❯❯❯ aws cloudformation validate-template --template-body file://templates/template.yaml
{
    "Parameters": []
}

(py36-default) ❯❯❯ echo $?
0

(py36-default) ❯❯❯ sceptre validate  prod/test-stack.yaml
Template prod/test-stack is valid. Template details:
{'Parameters': []}

(py36-default) ❯❯❯ sceptre --output yaml generate prod/test-stack.yaml
An error occured whilst writing the YAML object.
--- []
```

**Comment by zaro0508** [comment]
Verified that PR https://github.com/cloudreach/sceptre/pull/685 fixes this issue.

**Comment by craighurley** [comment]
I might be worth adding that I have similar issues generating json.  Generating text and yaml works fine.

I have **not** verififed if #685 fixes this issue.

Let me know if I should create a separate issue for this?

## Generate commands

```
$ sceptre --debug --output json generate dev
[2019-05-19 11:46:20] - Adding yaml constructors for the entry point groups ['sceptre.hooks', 'sceptre.resolvers']
[2019-05-19 11:46:20] - Added constructor for <class 'sceptre.hooks.asg_scaling_processes.ASGScalingProcesses'> with node tag !asg_scheduled_actions
[2019-05-19 11:46:20] - Added constructor for <class 'sceptre.hooks.cmd.Cmd'> with node tag !cmd
[2019-05-19 11:46:20] - Added constructor for <class 'sceptre.resolvers.environment_variable.EnvironmentVariable'> with node tag !environment_variable
[2019-05-19 11:46:20] - Added constructor for <class 'sceptre.resolvers.file_contents.FileContents'> with node tag !file_contents
[2019-05-19 11:46:20] - Added constructor for <class 'sceptre.resolvers.stack_output.StackOutput'> with node tag !stack_output
[2019-05-19 11:46:20] - Added constructor for <class 'sceptre.resolvers.stack_output.StackOutputExternal'> with node tag !stack_output_external
[2019-05-19 11:46:20] - Reading in 'dev/a/config.yaml' files...
[2019-05-19 11:46:20] - Config: {'project_path': '/Users/craighurley/Projects/aws/stack-test', 'stack_group_path': 'dev/a', 'environment': 'dev', 'project_code': 'testing', 'region': 'eu-west-1'}
[2019-05-19 11:46:20] - Reading in 'dev/a/stack.yaml' files...
[2019-05-19 11:46:20] - Config: {'project_path': '/Users/craighurley/Projects/aws/stack-test', 'stack_group_path': 'dev/a', 'environment': 'dev', 'project_code': 'testing', 'region': 'eu-west-1', 'template_path': 'stack.yaml'}
[2019-05-19 11:46:20] - Generate edges for graph 
An error occured writing the JSON object.
[]
```

```
$ sceptre --debug --output text generate dev
[2019-05-19 11:46:14] - Adding yaml constructors for the entry point groups ['sceptre.hooks', 'sceptre.resolvers']
[2019-05-19 11:46:14] - Added constructor for <class 'sceptre.hooks.asg_scaling_processes.ASGScalingProcesses'> with node tag !asg_scheduled_actions
[2019-05-19 11:46:14] - Added constructor for <class 'sceptre.hooks.cmd.Cmd'> with node tag !cmd
[2019-05-19 11:46:14] - Added constructor for <class 'sceptre.resolvers.environment_variable.EnvironmentVariable'> with node tag !environment_variable
[2019-05-19 11:46:14] - Added constructor for <class 'sceptre.resolvers.file_contents.FileContents'> with node tag !file_contents
[2019-05-19 11:46:14] - Added constructor for <class 'sceptre.resolvers.stack_output.StackOutput'> with node tag !stack_output
[2019-05-19 11:46:14] - Added constructor for <class 'sceptre.resolvers.stack_output.StackOutputExternal'> with node tag !stack_output_external
[2019-05-19 11:46:14] - Reading in 'dev/a/config.yaml' files...
[2019-05-19 11:46:14] - Config: {'project_path': '/Users/craighurley/Projects/aws/stack-test', 'stack_group_path': 'dev/a', 'environment': 'dev', 'project_code': 'testing', 'region': 'eu-west-1'}
[2019-05-19 11:46:14] - Reading in 'dev/a/stack.yaml' files...
[2019-05-19 11:46:14] - Config: {'project_path': '/Users/craighurley/Projects/aws/stack-test', 'stack_group_path': 'dev/a', 'environment': 'dev', 'project_code': 'testing', 'region': 'eu-west-1', 'template_path': 'stack.yaml'}
[2019-05-19 11:46:14] - Generate edges for graph 
["AWSTemplateFormatVersion: '2010-09-09'\nResources:\n  LogGroup:\n    Type: AWS::Logs::LogGroup\n    Properties:\n    RetentionInDays: 7\n"]
```

```
$ sceptre --debug --output yaml generate dev
[2019-05-19 11:46:17] - Adding yaml constructors for the entry point groups ['sceptre.hooks', 'sceptre.resolvers']
[2019-05-19 11:46:17] - Added constructor for <class 'sceptre.hooks.asg_scaling_processes.ASGScalingProcesses'> with node tag !asg_scheduled_actions
[2019-05-19 11:46:17] - Added constructor for <class 'sceptre.hooks.cmd.Cmd'> with node tag !cmd
[2019-05-19 11:46:17] - Added constructor for <class 'sceptre.resolvers.environment_variable.EnvironmentVariable'> with node tag !environment_variable
[2019-05-19 11:46:17] - Added constructor for <class 'sceptre.resolvers.file_contents.FileContents'> with node tag !file_contents
[2019-05-19 11:46:17] - Added constructor for <class 'sceptre.resolvers.stack_output.StackOutput'> with node tag !stack_output
[2019-05-19 11:46:17] - Added constructor for <class 'sceptre.resolvers.stack_output.StackOutputExternal'> with node tag !stack_output_external
[2019-05-19 11:46:17] - Reading in 'dev/a/config.yaml' files...
[2019-05-19 11:46:17] - Config: {'project_path': '/Users/craighurley/Projects/aws/stack-test', 'stack_group_path': 'dev/a', 'environment': 'dev', 'project_code': 'testing', 'region': 'eu-west-1'}
[2019-05-19 11:46:17] - Reading in 'dev/a/stack.yaml' files...
[2019-05-19 11:46:17] - Config: {'project_path': '/Users/craighurley/Projects/aws/stack-test', 'stack_group_path': 'dev/a', 'environment': 'dev', 'project_code': 'testing', 'region': 'eu-west-1', 'template_path': 'stack.yaml'}
[2019-05-19 11:46:17] - Generate edges for graph 
---
- AWSTemplateFormatVersion: '2010-09-09'
  Resources:
    LogGroup:
      Properties: null
      RetentionInDays: 7
      Type: AWS::Logs::LogGroup
```

## Project layout and files

```
$ tree 
.
├── config
│   └── dev
│       ├── a
│       │   └── stack.yaml
│       └── config.yaml
└── templates
    └── stack.yaml

4 directories, 3 files
```

```
$ cat config/dev/a/stack.yaml 
template_path: stack.yaml
```

```
$ cat templates/stack.yaml 
AWSTemplateFormatVersion: '2010-09-09'
Resources:
  LogGroup:
    Type: AWS::Logs::LogGroup
    Properties:
    RetentionInDays: 7
```

## Versions

```
$ sceptre --version
Sceptre, version 2.1.3
```

```
$ python --version
Python 3.6.8
```

```
$ pip list
Package            Version 
------------------ --------
aws-sam-translator 1.11.0  
boto3              1.9.151 
botocore           1.12.151
certifi            2019.3.9
cfn-lint           0.20.3  
chardet            3.0.4   
Click              7.0     
colorama           0.3.9   
decorator          4.4.0   
docutils           0.14    
idna               2.8     
Jinja2             2.10.1  
jmespath           0.9.4   
jsonpatch          1.23    
jsonpointer        2.0     
jsonschema         2.6.0   
MarkupSafe         1.1.1   
networkx           2.1     
packaging          16.8    
pathspec           0.5.9   
pip                19.1.1  
pyparsing          2.4.0   
python-dateutil    2.8.0   
PyYAML             5.1     
requests           2.21.0  
s3transfer         0.2.0   
sceptre            2.1.3   
setuptools         40.6.2  
six                1.12.0  
urllib3            1.24.3  
yamllint           1.15.0  
```

**Comment by ngfgrant** [comment]
@jk2l what exact version of sceptre2 are you using?

**Comment by jk2l** [comment]
@ngfgrant 2.1.5

if you look into craighurley's comment above 

```
$ sceptre --debug --output yaml generate dev
[2019-05-19 11:46:17] - Adding yaml constructors for the entry point groups ['sceptre.hooks', 'sceptre.resolvers']
[2019-05-19 11:46:17] - Added constructor for <class 'sceptre.hooks.asg_scaling_processes.ASGScalingProcesses'> with node tag !asg_scheduled_actions
[2019-05-19 11:46:17] - Added constructor for <class 'sceptre.hooks.cmd.Cmd'> with node tag !cmd
[2019-05-19 11:46:17] - Added constructor for <class 'sceptre.resolvers.environment_variable.EnvironmentVariable'> with node tag !environment_variable
[2019-05-19 11:46:17] - Added constructor for <class 'sceptre.resolvers.file_contents.FileContents'> with node tag !file_contents
[2019-05-19 11:46:17] - Added constructor for <class 'sceptre.resolvers.stack_output.StackOutput'> with node tag !stack_output
[2019-05-19 11:46:17] - Added constructor for <class 'sceptre.resolvers.stack_output.StackOutputExternal'> with node tag !stack_output_external
[2019-05-19 11:46:17] - Reading in 'dev/a/config.yaml' files...
[2019-05-19 11:46:17] - Config: {'project_path': '/Users/craighurley/Projects/aws/stack-test', 'stack_group_path': 'dev/a', 'environment': 'dev', 'project_code': 'testing', 'region': 'eu-west-1'}
[2019-05-19 11:46:17] - Reading in 'dev/a/stack.yaml' files...
[2019-05-19 11:46:17] - Config: {'project_path': '/Users/craighurley/Projects/aws/stack-test', 'stack_group_path': 'dev/a', 'environment': 'dev', 'project_code': 'testing', 'region': 'eu-west-1', 'template_path': 'stack.yaml'}
[2019-05-19 11:46:17] - Generate edges for graph 
---
- AWSTemplateFormatVersion: '2010-09-09'
  Resources:
    LogGroup:
      Properties: null
      RetentionInDays: 7
      Type: AWS::Logs::LogGroup
```

in the output, you can see "-" infront of "AWSTemplateFormatVersion". this mean it is array. this array cfn template cannot be directly used by CloudFormation

**Comment by ngfgrant** [comment]
Ok I think there are two separate issues here. The original issues was relating to specific cfn tags. I think it would be worth opening up a new issue for what you are describing @jk2l

**Comment by zaro0508** [comment]
I think the original problem identified in this issue is fixed?  I've been using sceptre version 2.1.4. 
 Been using the generate command with complex cfn templates with jinja and it's been working great for me.  I do see the same issue as @jk2l but that is a different issue.  I think we can safely close this one now?

**Comment by ngfgrant** [comment]
Yep the original issue is fixed @zaro0508 :)
